### PR TITLE
IGNITE-14920: Implement Spring Sessions Using Ignite As Backing Store

### PR DESCRIPTION
**Pull request by atris**
This PR implements functionality for storing Spring sessions in Ignite.

Originally written by Semyon Danilov.

### CODE REVIEW

**Comment by atris** [comment]
@SammyVimes Please review

File: modules/spring-sessions/pom.xml [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            <groupId>org.apache.ignite</groupId>
+            <artifactId>ignite-core</artifactId>
+            <version>${ignite.version}</version>
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-core</artifactId>
+            <version>5.3.8</version>
```
Comment (SammyVimes) [inline modules/spring-sessions/pom.xml:26]: please extract versions into variables

File: modules/spring-sessions/pom.xml [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        <dependency>
+            <groupId>javax.annotation</groupId>
+            <artifactId>javax.annotation-api</artifactId>
+            <version>1.3.2</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.ignite</groupId>
+            <artifactId>ignite-core</artifactId>
+            <version>2.11.0-SNAPSHOT</version>
```
Comment (SammyVimes) [inline modules/spring-sessions/pom.xml:65]: I don't think we should use SNAPSHOT versions here

File: modules/spring-sessions/pom.xml [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            <artifactId>mockito-core</artifactId>
+            <version>2.22.0</version>
+            <scope>test</scope>
+        </dependency>
+
+
+    </dependencies>
+
+
+</project>
```
Comment (SammyVimes) [inline modules/spring-sessions/pom.xml:134]: Missing newline

File: modules/spring-sessions/src/main/java/org/apache/ignite/spring/sessions/EnableIgniteHttpSession.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+    /**
+     * Save mode for the session. The default is {@link SaveMode#ON_SET_ATTRIBUTE}, which
+     * only saves changes made to session.
+     * @return the save mode
+     * @since 2.2.0
+     */
+    SaveMode saveMode() default SaveMode.ON_SET_ATTRIBUTE;
+
+}
```
Comment (SammyVimes) [inline modules/spring-sessions/src/main/java/org/apache/ignite/spring/sessions/EnableIgniteHttpSession.java:98]: Missing newline

File: modules/spring-sessions/src/test/java/org/apache/ignite/spring/sessions/ClientServerIgniteIndexedSessionRepositoryITests.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Integration tests for {@link IgniteIndexedSessionRepository} using client-server
+ * topology.
+ */
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration
+@WebAppConfiguration
+class ClientServerIgniteIndexedSessionRepositoryITests extends AbstractIgniteIndexedSessionRepositoryITests {
+
+    private static GenericContainer container = new GenericContainer<>("apacheignite/ignite:2.9.0")
```
Comment (SammyVimes) [inline modules/spring-sessions/src/test/java/org/apache/ignite/spring/sessions/ClientServerIgniteIndexedSessionRepositoryITests.java:30]: I doubt that we can use test-containers with our current TeamCity configuration. Please verify it by running team city tests for ignite-extensions

**Review by SammyVimes — CHANGES_REQUESTED** [review]
It looks ok in general, but it doesn't meet Apache Ignite code style requirements:
1) Missing newlines at the end of the files
2) Override should be on the same line with the method declaration (e.g. `@Override public static...`
3) Each method and field should have a JavaDoc
4) Braces should be ommited for if/for/while operator bodies if the body is only one line of code

File: modules/spring-sessions/src/test/java/org/apache/ignite/spring/sessions/ClientServerIgniteIndexedSessionRepositoryITests.java (hunk 5 above)
Comment (atris) [reply modules/spring-sessions/src/test/java/org/apache/ignite/spring/sessions/ClientServerIgniteIndexedSessionRepositoryITests.java]: Agreed, deleted these tests

**Review by SammyVimes — APPROVED** [review]
Looks good to me

File: modules/spring-session-ext/pom.xml [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>javax.annotation</groupId>
+            <artifactId>javax.annotation-api</artifactId>
+            <version>1.3.2</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.ignite</groupId>
```
Comment (SammyVimes) [inline modules/spring-session-ext/pom.xml:63]: Odd dependency: ignite-core is already defined

File: modules/spring-session-ext/pom.xml [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            <artifactId>assertj-core</artifactId>
+            <version>3.20.0</version>
+            <scope>test</scope>
+        </dependency>
+
+        <!-- https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-api -->
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>5.7.2</version>
```
Comment (SammyVimes) [inline modules/spring-session-ext/pom.xml:81]: all dependencies' versions should be extracted to variables

File: modules/spring-session-ext/pom.xml [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-web</artifactId>
+            <version>${spring.version.for.sessions}</version>
+        </dependency>
+
+        <!-- https://mvnrepository.com/artifact/org.apache.ignite/ignite-indexing -->
+        <dependency>
+            <groupId>org.apache.ignite</groupId>
+            <artifactId>ignite-indexing</artifactId>
+            <version>2.10.0</version>
```
Comment (SammyVimes) [inline modules/spring-session-ext/pom.xml:119]: use ignite.version variable

File: modules/spring-session-ext/pom.xml [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            <groupId>org.apache.ignite</groupId>
+            <artifactId>ignite-core</artifactId>
+            <version>${ignite.version}</version>
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-core</artifactId>
+            <version>${spring.version.for.sessions}</version>
```
Comment (SammyVimes) [inline modules/spring-session-ext/pom.xml:26]: should be just spring.version or something like that. That stands for all of the variables' names.

**Review by SammyVimes — CHANGES_REQUESTED** [review]

File: modules/spring-session-ext/src/test/java/org/apache/ignite/spring/sessions/IgniteIndexedSessionRepositoryTests.java [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+
+/**
+ * Tests for {@link IgniteIndexedSessionRepository}.
+ */
+class IgniteIndexedSessionRepositoryTests {
```
Comment (alamar) [inline modules/spring-session-ext/src/test/java/org/apache/ignite/spring/sessions/IgniteIndexedSessionRepositoryTests.java:61]: There should be no blank line after class start. There should, however, be a javadoc (empty is OK) before every field and method declared. The file should probably be called FooTest and not FooTests. Test files should be added to suite. I will assist in adding suite to TC once it is there.

Comment (atris) [reply modules/spring-session-ext/src/test/java/org/apache/ignite/spring/sessions/IgniteIndexedSessionRepositoryTests.java]: Fixed, thanks

**Comment by atris** [comment]
@alamar Updated, please see

**Comment by SammyVimes** [comment]
@alamar The build and tests are ok

https://ci.ignite.apache.org/buildConfiguration/IgniteExtensions_Tests_SpringSession/6057126?buildTab=overview

**Comment by atris** [comment]
@alamar @SammyVimes @ibessonov This PR was merged without squashing commits and has populated the master history with excessive commits. Can we please revert and squash merge?

**Pull request merged**